### PR TITLE
Add ability to enable/disable preview features

### PR DESCRIPTION
**Pull request by halspang**
# Description
The dapr runtime configuration now includes a FeatureSpec that
allows for features declared enabled/disabled. The section of
the runtime that calls into/handles the feature must read this
spec to determine if the feature is active or not. Currently,
only Actor Reentrancy is defined.

## Issue reference

Please reference the issue this PR will close: #3048 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/config/configuration.go [hunk 1]
```diff
@@ -84,6 +85,7 @@ type ConfigurationSpec struct {
 	MetricSpec        MetricSpec        `json:"metric,omitempty" yaml:"metric,omitempty"`
 	Secrets           SecretsSpec       `json:"secrets,omitempty" yaml:"secrets,omitempty"`
 	AccessControlSpec AccessControlSpec `json:"accessControl,omitempty" yaml:"accessControl,omitempty"`
+	FeatureSpec       []FeatureSpec     `json:"features,omitempty" yaml:"features,omitempty"`
```
Comment (artursouza) [inline pkg/config/configuration.go:88]: Just `Features`, similar to `Secrets`. We should probably remove the `Spec` suffix from the other attributes too.

File: pkg/config/configuration.go [hunk 2]
```diff
@@ -170,6 +172,12 @@ type SpiffeID struct {
 	AppID       string
 }
 
+// FeatureSpec defines which preview features are enabled
+type FeatureSpec struct {
+	FeatureName string `json:"name" yaml:"name"`
```
Comment (artursouza) [inline pkg/config/configuration.go:177]: Just `Name`.

**Review by artursouza — COMMENTED** [review]
Please, modify https://github.com/dapr/dapr/blob/master/charts/dapr/crds/configuration.yaml too

E2E tests would probably be added once we test reentrancy.

File: pkg/config/configuration.go (hunk 1 above)
Comment (halspang) [reply pkg/config/configuration.go:88]: Done and I agree they should be normalized. I can through that in here but would prefer a separate PR for it.

File: pkg/config/configuration.go (hunk 2 above)
Comment (halspang) [reply pkg/config/configuration.go:177]: Done.

**Comment by artursouza** [comment]
@halspang Please, resolve conflicts.

**Comment by yaron2** [comment]
I don't see the required changes on the CRD structure: https://github.com/dapr/dapr/blob/9e32048cf5b84fd33834cbc48f8f94a72554ac74/pkg/apis/configuration/v1alpha1/types.go#L20

Once the above is modified, please re-generate the deepcopy files using `make code-generate`.

File: pkg/config/configuration.go [hunk 3]
```diff
@@ -43,6 +43,7 @@ const (
 	SpiffeIDPrefix      = "spiffe://"
 	HTTPProtocol        = "http"
 	GRPCProtocol        = "grpc"
+	ActorRentrancy      = "Actor.Reentrancy"
```
Comment (artursouza) [inline pkg/config/configuration.go:46]: Please, create some sort of enum with features and make them strongly typed. For example: https://www.ribice.ba/golang-enums/

**Review by artursouza — CHANGES_REQUESTED** [review]
Just one minor thing.

**Review by artursouza — APPROVED** [review]

**Comment by artursouza** [comment]
@yaron2 Can we merge this?

**Comment by artursouza** [comment]
@yaron2 I've approved this PR since I saw the CRD changes. If you think it needs anything else, please, ask @halspang to fix in a new PR.

**Pull request merged**